### PR TITLE
th/dbus_cache_properties

### PR DESCRIPTION
**Pull request by thom311**
Cache property GVariants for D-Bus and make more use of `CList` to track devices in manager and `NMWifiAP`

### CODE REVIEW

**Comment by thom311** [comment]
@balrog-kun, some changes to IWD code here. Does it look OK?

File: src/nm-manager.c [hunk 1]
```diff
@@ -799,18 +799,20 @@ nm_manager_get_activatable_connections (NMManager *manager, guint *out_len, gboo
...
+	NMManagerPrivate *priv = NM_MANAGER_GET_PRIVATE (self);
 	NMActiveConnection *ac;
 
-	nm_assert (path);
+	ac = (NMActiveConnection *) nm_dbus_manager_lookup_object (nm_dbus_object_get_manager (NM_DBUS_OBJECT (self)),
+	                                                           path);
+	if (   !ac
+	    || !NM_IS_ACTIVE_CONNECTION (ac)
+	    || c_list_is_empty (&ac->active_connections_lst))
+		return NULL;
```
[the remark this answers is not in the document]
Comment (thom311) [reply src/nm-manager.c]: In general, I am in favor of assertions, but in this case, I think it's too much. Note already before, the code allowed for the possibility that the path may not exist. The possibility of a non-existing path, means the caller apparently has no control about the path he looks up, so asserting against it is wrong. Maybe in this case, for `active_connection_get_by_path()` you are right. But for example `nm_wifi_ap_lookup_for_device()` does something similar, and that is called with a path provided by the user (untrusted).

File: src/nm-manager.c [hunk 2]
```diff
@@ -933,11 +935,15 @@ nm_manager_get_device_by_path (NMManager *self, const char *path)
...
-		if (!strcmp (nm_dbus_object_get_path (NM_DBUS_OBJECT (device)), path))
-			return device;
-	}
-	return NULL;
+	device = (NMDevice *) nm_dbus_manager_lookup_object (nm_dbus_object_get_manager (NM_DBUS_OBJECT (self)),
+	                                                     path);
+	if (   !device
+	    || !NM_IS_DEVICE (device)
+	    || c_list_is_empty (&device->devices_lst))
+		return NULL;
```
[the remark this answers is not in the document]
Comment (thom311) [reply src/nm-manager.c]: See how `nm_checkpoint_manager_create()` looks paths up from untrusted source.

File: src/devices/nm-device.h [hunk 3]
```diff
@@ -174,6 +174,7 @@ struct _NMDevicePrivate;
 struct _NMDevice {
 	NMDBusObject parent;
 	struct _NMDevicePrivate *_priv;
+	CList devices_lst;
```
[the remark this answers is not in the document]
Comment (thom311) [reply src/devices/nm-device.h]: Extended commit message. Ok?

**Comment by bengal** [comment]
LGTM now

**Comment by thom311** [comment]
Thanks. Merged

**Pull request merged**